[refactor] Replace conditional ladder in requests utils

nikochiko opened the pull request:
As a Google Code-In task.

Replace conditional ladder in requests utils, with a single `requests.request` call.

vkartik97 commented on evalai/utils/requests.py:
```diff
@@ -90,19 +62,12 @@ def make_request(path, method, files=None, data=None):
-        pass
+    # TODO: Add support for PUT request
+    # TODO: Add support for PATCH request
+    # TODO: Add support for DELETE request
```
Please remove these TODOs as the `requests.request` will implement this as well

nikochiko replied:
Okay. That's done! :D :+1:

nikochiko commented:
@vkartik97 @RishabhJain2018 Shall I explicitly remove support for PUT, PATCH, DELETE requests? 
I realized in the last comment that it accepts all methods and that could give unexpected results.

vkartik97 commented:
> I realized in the last comment that it accepts all methods and that could give unexpected results.

Can you please provide detailed information regarding this?

nikochiko commented on evalai/utils/requests.py:
```diff
@@ -13,73 +12,46 @@
                 style(
-                    "\nCould not establish a connection to EvalAI."
-                    " Please check the Host URL.\n",
+                    "Use `evalai challenges` to fetch the active challenges."
```
@vkartik97 As in, it wasn't there in the code previously. So, it might not make sense to show this message in that case (for example for DELETE method). Also, I'm not sure if we have the support for this in the backend.

nikochiko replied:
@vkartik97 It would be better to have some error messages / blocking requests due to security reasons (that can be possible with a custom script, but we can still make a warning block here). For reference: https://www.owasp.org/index.php/Test_HTTP_Methods_(OTG-CONFIG-006)

vkartik97 replied:
I am thinking of having something like:
```py
if method in [PUT,PATCH, DELETE] then:
    return or/and throw error
else:
.
.
.

```
What are your opinions on this?

nikochiko replied:
Great, looks good to me.
I checked the [exceptions list](https://requests.readthedocs.io/en/master/_modules/requests/exceptions/) provided by requests package but it doesn't include an `InvalidMethodError`. So I was thinking we could `echo` a message saying the method is unsupported and then perform a `sys.exit`.
That is:
```python
if method in ["PUT", "PATCH", "DELETE"]:
  echo("Method not supported ...")
  sys.exit(1)
else:
.
.
.
```

vkartik97 replied:
Instead of `echo` and `exit`, what are opinions on using `raise Exception('{METHOD} not supported by make_request')`? The reason for using exceptions is that the method `make_request` should have a behavior to inform the user/programmer that a exceptional behavior has occurred, here it is `"PUT", "PATCH", "DELETE" is not supported`.
Please let me know if this doesn't seem correct way.

nikochiko replied:
@vkartik97 I think that is a valid point. I agree raising an Exception will be better :+1:
I will update the PR with the same.

nikochiko commented on evalai/utils/requests.py:
```diff
@@ -11,75 +10,58 @@
 
 
 def make_request(path, method, files=None, data=None):
+    if method in  ["PUT", "PATCH", "DELETE"]:
```
@vkartik97 looks good?

vkartik97 commented on evalai/utils/requests.py:
```diff
@@ -11,75 +10,58 @@
+                fg="red",
+            )
+        )
+        sys.exit(1)
```
Don't echo, throw error in this case

nikochiko replied:
Okay. `ValueError`?

vkartik97 replied:
Why value error?

nikochiko replied:
I think that can be appropriate because the `method` parameter got an invalid value as an argument. 
What do you say?

nikochiko replied:
@vkartik97 It can more appropriately be called 'UnsupportedMethodError' or 'InvalidArgumentError' but since we don't have custom exception types, 'ValueError' can be used (as the InvalidArgumentError can be considered a form of ValueError).

Otherwise, I think just calling it an `Exception` would be okay and clear all confusion. 
We can also create custom exception types if it can be used in multiple cases. This can be the best thing to do but I'm not sure if it's feasible if it only has one use case. 
What is your opinion on this?

Ram81 commented on evalai/utils/requests.py:
```diff
@@ -11,75 +10,51 @@
-                    "\nCould not establish a connection to EvalAI."
-                    " Please check the Host URL.\n",
-                    bold=True,
+                    "\nError: {}\n".format(response.json().get("error")),
```
@nikochiko can we add some description of the error like it was there before `Could not establish a connection to EvalAI`

nikochiko replied:
@Ram81 That error message is more appropriate for the RequestException where it is being shown, see [this line](https://github.com/Cloud-CV/evalai-cli/pull/237/files#diff-e7df4381b09aa49cce42fde57be6249dR49). As this handling part is for HTTP errors, the error types may vary and therefore I have left this generalized message.

vkartik97 replied:
@Ram81 @nikochiko 
How about
```
Could not establish a connection to EvalAI with error: {ERROR}
```

nikochiko replied:
@vkartik97 But it could also be a HTTP-500 error / HTTP-40X error when the connection was established but the request couldn't be completed. The reference to a connection issue could be misleading in such a case.

nikochiko replied:
@Ram81 I have not made any changes in the behavior of this function, I think the GitHub differences notation is a bit misleading, the error message being shown in red was for the handling of `requests.exceptions.RequestException` and the one in green that is being handled here is for HTTPError.

vkartik97 replied:
Pardon me, I didnt explain better.
I meant that 
```
except requests.exceptions.HTTPError as e:
.
.
.
.
          echo("Could not establish a connection to EvalAI with error: {0}".format(e))
```
The error should give error that `requests` encountered. This will give context to the user of the type of error.

nikochiko replied:
@vkartik97 Okay, that looks good, Actually we use the same error message for `requests.exceptions.RequestException`, so I think we can merge both. We also validate_token in between if a known error is thrown by the server, So, how about this to combine:
```python
except requests.exceptions.RequestException as e:
    if isinstance(e, requests.exceptions.HTTPError) and response.status_code in EVALAI_ERROR_CODES:
        validate_token(response.json())
        e = response.json().get("error")  # In this case, the error message is returned by the server
    echo("Could not establish a connection to EvalAI with error: {}".format(e))
``` 
This will shorten and generalize the error handling. Is this good?

nikochiko replied:
I think I had misunderstood it, I will change it according to the syntax mentioned!

Ram81 commented on evalai/utils/requests.py:
```diff
@@ -11,75 +10,51 @@
+
     url = "{}{}".format(get_host_url(), path)
     headers = get_request_header()
+    if method == "POST":
```
Add a new line before this

nikochiko replied:
Done! 😄

nikochiko commented:
@vkartik97 @Ram81 I have made the changes. Please see if it looks good.

nikochiko commented:
I have changed the handling mechanism a bit, but the behavior of the function is the same, and with the suggested additions.
All exceptions caused by `requests` have been generalized into a single block, and all error messages have the same template.
This has also shortened the code and removed the multiple `except` calls.
Please review.

nikochiko commented:
@Ram81 Can I get an approval on the GCI page?
